diff -Naur bdwgc.orig/include/gc_pthread_redirects.h bdwgc/include/gc_pthread_redirects.h
--- bdwgc.orig/include/gc_pthread_redirects.h	2010-09-02 08:05:10.000000000 +0200
+++ bdwgc/include/gc_pthread_redirects.h	2011-03-05 20:38:26.000000000 +0100
@@ -60,6 +60,10 @@
 # endif
 #endif
 
+#if defined(PLATFORM_ANDROID)
+#  undef GC_PTHREAD_EXIT_ATTRIBUTE
+#endif
+
 #ifdef GC_PTHREAD_EXIT_ATTRIBUTE
   GC_API int GC_pthread_cancel(pthread_t);
   GC_API void GC_pthread_exit(void *) GC_PTHREAD_EXIT_ATTRIBUTE;
