diff -Naur gc-7.5.0-27may2016.orig/alloc.c gc-7.5.0-27may2016/alloc.c
--- gc-7.5.0-27may2016.orig/alloc.c	2016-05-27 09:11:49.882369814 +0200
+++ gc-7.5.0-27may2016/alloc.c	2016-05-27 09:20:34.560136370 +0200
@@ -573,6 +573,40 @@
     RESTORE_CANCEL(cancel_state);
 }
 
+/*---------------------------------------------------------------------*/
+/*    Bigloo profiling 13Apr2003: start                                */
+/*---------------------------------------------------------------------*/
+static void (*GC_gcollect_hook)() = 0;
+void GC_add_gc_hook( void (*f)() ) {
+   GC_gcollect_hook = f;
+}
+/*---------------------------------------------------------------------*/
+/*    Bigloo profiling 13Apr2003: stop                                 */
+/*---------------------------------------------------------------------*/
+
+/*---------------------------------------------------------------------*/
+/*    Bigloo start                                                     */
+/*---------------------------------------------------------------------*/
+static long allocated_bytes = 0;
+
+GC_get_allocated_bytes() {
+   return allocated_bytes;
+}
+
+void GC_reset_allocated_bytes() {
+   GC_gcollect();
+   allocated_bytes = 0;
+}
+
+#ifdef KEEP_BACK_PTRS
+int GC_heap_info_available = 1;
+#else
+int GC_heap_info_available = 0;
+#endif
+/*---------------------------------------------------------------------*/
+/*    Bigloo stop                                                      */
+/*---------------------------------------------------------------------*/
+
 GC_INNER void (*GC_check_heap)(void) = 0;
 GC_INNER void (*GC_print_all_smashed)(void) = 0;
 
@@ -912,6 +946,16 @@
         GC_check_tls();
 #   endif
 
+/*---------------------------------------------------------------------*/
+/*    Bigloo start                                                     */
+/*---------------------------------------------------------------------*/
+#   ifdef KEEP_BACK_PTRS
+    allocated_bytes += GC_get_bytes_since_gc();
+#   endif
+/*---------------------------------------------------------------------*/
+/*    Bigloo stop                                                      */
+/*---------------------------------------------------------------------*/
+
 #   ifndef SMALL_CONFIG
       if (GC_print_stats)
         GET_TIME(start_time);
@@ -1011,6 +1055,17 @@
                           COMMA_IF_USE_MUNMAP((unsigned long)
                                               GC_unmapped_bytes));
 
+/*---------------------------------------------------------------------*/
+/*    Bigloo profiling 13Apr2003: start                                */
+/*---------------------------------------------------------------------*/
+    if( GC_gcollect_hook )
+       GC_gcollect_hook( GC_heapsize,
+			 (unsigned long)(GC_atomic_in_use)+
+			 (unsigned long)(GC_composite_in_use) );
+/*---------------------------------------------------------------------*/
+/*    Bigloo profiling 13Apr2003: start                                */
+/*---------------------------------------------------------------------*/
+    
     /* Reset or increment counters for next cycle */
     GC_n_attempts = 0;
     GC_is_full_gc = FALSE;
--- gc-7.5.0-27may2016.orig/Makefile.am	2016-05-27 09:11:49.882369814 +0200
+++ gc-7.5.0-27may2016/Makefile.am	2016-05-27 09:16:03.429158747 +0200
@@ -53,12 +53,24 @@
 libgc_la_SOURCES = extra/gc.c
 else
 EXTRA_DIST += extra/gc.c
+#*---------------------------------------------------------------------*/
+#*    Bigloo start (16 mars 2012)                                      */
+#*---------------------------------------------------------------------*/
+#libgc_la_SOURCES = \
+#    allchblk.c alloc.c backgraph.c blacklst.c checksums.c dbg_mlc.c \
+#    dyn_load.c finalize.c gc_dlopen.c gcj_mlc.c headers.c \
+#    mach_dep.c malloc.c mallocx.c mark.c mark_rts.c misc.c new_hblk.c \
+#    obj_map.c os_dep.c pcr_interface.c ptr_chck.c real_malloc.c reclaim.c \
+#    specific.c stubborn.c thread_local_alloc.c typd_mlc.c
 libgc_la_SOURCES = \
-    allchblk.c alloc.c blacklst.c checksums.c dbg_mlc.c \
+    allchblk.c alloc.c inline_alloc.c gc_intf.c blacklst.c checksums.c dbg_mlc.c \
     dyn_load.c finalize.c gc_dlopen.c gcj_mlc.c headers.c \
     mach_dep.c malloc.c mallocx.c mark.c mark_rts.c misc.c new_hblk.c \
     obj_map.c os_dep.c pcr_interface.c ptr_chck.c real_malloc.c reclaim.c \
     specific.c stubborn.c thread_local_alloc.c typd_mlc.c
+#*---------------------------------------------------------------------*/
+#*    Bigloo stop (16 mars 2012)                                       */
+#*---------------------------------------------------------------------*/
 
 # C Library: Architecture Dependent
 # ---------------------------------
diff -Naur gc-7.5.0-27may2016.orig/Makefile.in gc-7.5.0-27may2016/Makefile.in
--- gc-7.5.0-27may2016.orig/Makefile.in	2016-05-27 09:12:29.382804599 +0200
+++ gc-7.5.0-27may2016/Makefile.in	2016-05-27 09:19:17.751293316 +0200
@@ -233,14 +233,20 @@
 	$(LIBTOOLFLAGS) --mode=link $(CCLD) $(AM_CFLAGS) $(CFLAGS) \
 	$(libcord_la_LDFLAGS) $(LDFLAGS) -o $@
 am__DEPENDENCIES_1 =
-am__libgc_la_SOURCES_DIST = allchblk.c alloc.c blacklst.c checksums.c \
-	dbg_mlc.c dyn_load.c finalize.c gc_dlopen.c gcj_mlc.c \
-	headers.c mach_dep.c malloc.c mallocx.c mark.c mark_rts.c \
-	misc.c new_hblk.c obj_map.c os_dep.c pcr_interface.c \
-	ptr_chck.c real_malloc.c reclaim.c specific.c stubborn.c \
-	thread_local_alloc.c typd_mlc.c win32_threads.c \
+#*---------------------------------------------------------------------*/
+#*    Bigloo start (16 mars 2012)                                      */
+#*---------------------------------------------------------------------*/
+am__libgc_la_SOURCES_DIST = allchblk.c alloc.c inline_alloc.c gc_intf.c blacklst.c \
+	checksums.c dbg_mlc.c dyn_load.c finalize.c gc_dlopen.c \
+	gcj_mlc.c headers.c mach_dep.c malloc.c mallocx.c mark.c \
+	mark_rts.c misc.c new_hblk.c obj_map.c os_dep.c \
+	pcr_interface.c ptr_chck.c real_malloc.c reclaim.c specific.c \
+	stubborn.c thread_local_alloc.c typd_mlc.c win32_threads.c \
 	pthread_start.c pthread_support.c darwin_stop_world.c \
 	pthread_stop_world.c backgraph.c fnlz_mlc.c extra/gc.c
+#*---------------------------------------------------------------------*/
+#*    Bigloo stop (16 mars 2012)                                       */
+#*---------------------------------------------------------------------*/
 @SINGLE_GC_OBJ_FALSE@@WIN32_THREADS_TRUE@am__objects_1 =  \
 @SINGLE_GC_OBJ_FALSE@@WIN32_THREADS_TRUE@	win32_threads.lo
 @PTHREADS_TRUE@@SINGLE_GC_OBJ_FALSE@@WIN32_THREADS_FALSE@am__objects_2 = pthread_start.lo \
@@ -251,7 +257,10 @@
 @MAKE_BACK_GRAPH_TRUE@@SINGLE_GC_OBJ_FALSE@	backgraph.lo
 @ENABLE_DISCLAIM_TRUE@@SINGLE_GC_OBJ_FALSE@am__objects_6 =  \
 @ENABLE_DISCLAIM_TRUE@@SINGLE_GC_OBJ_FALSE@	fnlz_mlc.lo
-@SINGLE_GC_OBJ_FALSE@am_libgc_la_OBJECTS = allchblk.lo alloc.lo \
+#*---------------------------------------------------------------------*/
+#*    Bigloo start (16 mars 2012)                                      */
+#*---------------------------------------------------------------------*/
+@SINGLE_GC_OBJ_FALSE@am_libgc_la_OBJECTS = allchblk.lo alloc.lo inline_alloc.lo gc_intf.lo \
 @SINGLE_GC_OBJ_FALSE@	blacklst.lo checksums.lo dbg_mlc.lo \
 @SINGLE_GC_OBJ_FALSE@	dyn_load.lo finalize.lo gc_dlopen.lo \
 @SINGLE_GC_OBJ_FALSE@	gcj_mlc.lo headers.lo mach_dep.lo \
@@ -264,6 +273,9 @@
 @SINGLE_GC_OBJ_FALSE@	$(am__objects_2) $(am__objects_3) \
 @SINGLE_GC_OBJ_FALSE@	$(am__objects_4) $(am__objects_5) \
 @SINGLE_GC_OBJ_FALSE@	$(am__objects_6)
+#*---------------------------------------------------------------------*/
+#*    Bigloo stop (16 mars 2012)                                       */
+#*---------------------------------------------------------------------*/
 @SINGLE_GC_OBJ_TRUE@am_libgc_la_OBJECTS = extra/gc.lo $(am__objects_1) \
 @SINGLE_GC_OBJ_TRUE@	$(am__objects_2) $(am__objects_3) \
 @SINGLE_GC_OBJ_TRUE@	$(am__objects_4) $(am__objects_5) \
@@ -1005,7 +1017,10 @@
 	libstaticrootslib2_test.la
 pkgconfigdir = $(libdir)/pkgconfig
 pkgconfig_DATA = bdw-gc.pc
-@SINGLE_GC_OBJ_FALSE@libgc_la_SOURCES = allchblk.c alloc.c blacklst.c \
+#*---------------------------------------------------------------------*/
+#*    Bigloo start (16 mars 2012)                                      */
+#*---------------------------------------------------------------------*/
+@SINGLE_GC_OBJ_FALSE@libgc_la_SOURCES = allchblk.c alloc.c inline_alloc.c gc_intf.c blacklst.c \
 @SINGLE_GC_OBJ_FALSE@	checksums.c dbg_mlc.c dyn_load.c \
 @SINGLE_GC_OBJ_FALSE@	finalize.c gc_dlopen.c gcj_mlc.c \
 @SINGLE_GC_OBJ_FALSE@	headers.c mach_dep.c malloc.c mallocx.c \
@@ -1017,6 +1032,9 @@
 @SINGLE_GC_OBJ_FALSE@	$(am__append_2) $(am__append_3) \
 @SINGLE_GC_OBJ_FALSE@	$(am__append_4) $(am__append_5) \
 @SINGLE_GC_OBJ_FALSE@	$(am__append_6) $(am__append_7)
+#*---------------------------------------------------------------------*/
+#*    Bigloo stop (16 mars 2012)                                       */
+#*---------------------------------------------------------------------*/
 @SINGLE_GC_OBJ_TRUE@libgc_la_SOURCES = extra/gc.c $(am__append_2) \
 @SINGLE_GC_OBJ_TRUE@	$(am__append_3) $(am__append_4) \
 @SINGLE_GC_OBJ_TRUE@	$(am__append_5) $(am__append_6) \
@@ -1434,6 +1452,13 @@
 
 @AMDEP_TRUE@@am__include@ @am__quote@./$(DEPDIR)/allchblk.Plo@am__quote@
 @AMDEP_TRUE@@am__include@ @am__quote@./$(DEPDIR)/alloc.Plo@am__quote@
+#*---------------------------------------------------------------------*/
+#*    Bigloo start (16 mars 2012)                                      */
+#*---------------------------------------------------------------------*/
+@AMDEP_TRUE@@am__include@ @am__quote@./$(DEPDIR)/inline_alloc.Plo@am__quote@
+#*---------------------------------------------------------------------*/
+#*    Bigloo stop (16 mars 2012)                                       */
+#*---------------------------------------------------------------------*/
 @AMDEP_TRUE@@am__include@ @am__quote@./$(DEPDIR)/backgraph.Plo@am__quote@
 @AMDEP_TRUE@@am__include@ @am__quote@./$(DEPDIR)/blacklst.Plo@am__quote@
 @AMDEP_TRUE@@am__include@ @am__quote@./$(DEPDIR)/checksums.Plo@am__quote@
