diff -Naur gc-7.5.0-27may2016.bgl/pthread_stop_world.c gc-7.5.0-27may2016/pthread_stop_world.c
--- gc-7.5.0-27may2016.bgl/pthread_stop_world.c	2016-05-27 09:11:49.000000000 +0200
+++ gc-7.5.0-27may2016/pthread_stop_world.c	2016-05-27 10:40:39.328891314 +0200
@@ -450,14 +450,16 @@
 #endif
 
 #ifdef USE_TKILL_ON_ANDROID
-  extern int tkill(pid_t tid, int sig); /* from sys/linux-unistd.h */
+  // extern int tkill(pid_t tid, int sig); /* from sys/linux-unistd.h */
+  extern int tgkill(int tgid, pid_t tid, int sig); /* from sys/linux-unistd.h */
 
   static int android_thread_kill(pid_t tid, int sig)
   {
     int ret;
     int old_errno = errno;
 
-    ret = tkill(tid, sig);
+    // ret = tkill(tid, sig);
+    ret = tgkill(-1,tid, sig);
     if (ret < 0) {
         ret = errno;
         errno = old_errno;
